python/uv/private/lock.bzl

```diff
@@ -15,22 +15,258 @@
 """A simple macro to lock the requirements.
 """
 
-load("@bazel_skylib//rules:write_file.bzl", "write_file")
-load("//python:py_binary.bzl", "py_binary")
+load("@bazel_skylib//rules:expand_template.bzl", "expand_template")
+load("//python:py_test.bzl", "py_test")
 load("//python/private:bzlmod_enabled.bzl", "BZLMOD_ENABLED")  # buildifier: disable=bzl-visibility
 
 visibility(["//..."])
 
-_REQUIREMENTS_TARGET_COMPATIBLE_WITH = select({
-    "@platforms//os:windows": ["@platforms//:incompatible"],
-    "//conditions:default": [],
-}) if BZLMOD_ENABLED else ["@platforms//:incompatible"]
+_uv_toolchain = Label("//python/uv:uv_toolchain_type")
+_py_toolchain = Label("//python:toolchain_type")
 
-def lock(*, name, srcs, out, upgrade = False, universal = True, args = [], **kwargs):
+_LockInfo = provider(
+    doc = "",
+    fields = {
+        "args": "",
+        "py": "",
+        "srcs": "",
+        "template": "",
+        "uv": "",
+    },
+)
+
+def _lock_impl(ctx):
+    args = ctx.attr.args
+    srcs = ctx.files.srcs
+    existing_output = ctx.files.existing_output
+    output = ctx.outputs.output
+
+    toolchain_info = ctx.toolchains[_uv_toolchain]
+    uv = toolchain_info.uv_toolchain_info.uv[DefaultInfo].files_to_run.executable
+
+    py_runtime = ctx.toolchains[_py_toolchain].py3_runtime
+
+    cmd = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.expand_template(
+        template = ctx.files._template[0],
+        substitutions = {
+            " uv ": " {} ".format(uv.path),
+            "bazel_out": output.path,
+            "replace": existing_output[0].path,
+        },
+        output = cmd,
+        is_executable = True,
+    )
+
+    run_args = []
+    lock_args = ctx.actions.args()
+    lock_args.add("--no-python-downloads")
+    run_args.append("--no-python-downloads")
+    lock_args.add("--no-cache")
+    run_args.append("--no-cache")
+    if ctx.attr.generate_hashes:
+        lock_args.add("--generate-hashes")
+        run_args.append("--generate-hashes")
+    if not ctx.attr.strip_extras:
+        lock_args.add("--no-strip-extras")
+        run_args.append("--no-strip-extras")
+
+    update_command = "bazel run //{}:{}".format(
+        ctx.label.package,
+        ctx.attr.update_target,
+    )
+    lock_args.add("--custom-compile-command", update_command)
+    run_args.extend(("--custom-compile-command", "'{}'".format(update_command)))
+
+    lock_args.add("--no-progress")
+    lock_args.add("--quiet")
+    lock_args.add("--python", py_runtime.interpreter)
+    lock_args.add_all(args)
+    lock_args.add_all(srcs)
+
+    ctx.actions.run(
+        executable = cmd,
```

Comment:
I think this will be cleaner if the build action calls uv directly without the shell script wrapper. The run-wrapper can generate the sh wrapper instead.

Reply:
We cannot do this because we need to copy the source copy of the requirements file to the ctx.outputs.output if it already exists. This is so that uv would know what versions to use for the final solution.

Reply:
Ah, yes we can, we just need to use `run_shell`.